easybuild/tools/run.py

```diff
@@ -205,17 +205,24 @@ def create_cmd_scripts(cmd_str, work_dir, env, tmpdir):
       and with the command in shell history;
     """
     # Save environment variables in env.sh which can be sourced to restore environment
-    full_env = os.environ.copy()
-    if env is not None:
-        full_env.update(env)
+    if env is None:
+        env = os.environ.copy()
+
     env_fp = os.path.join(tmpdir, 'env.sh')
     with open(env_fp, 'w') as fid:
         # excludes bash functions (environment variables ending with %)
-        fid.write('\n'.join(f'export {key}={shlex.quote(value)}' for key, value in sorted(full_env.items())
-                            if not key.endswith('%')))
+        fid.write('\n'.join(f'export {key}={shlex.quote(value)}' for key, value in sorted(env.items())
+                            if not key.endswith('%')) + '\n')
+
+        # unset environment variables in current environment if they're not defined in environment used to run command
+        fid.write('\n'.join(f'unset {key}' for key in os.environ if key not in env and not key.endswith('%')))
```

Comment:
hm, this isn't really correct though? When a user sources this, they will likely have other variables set. 
It should unset everything the user has as the point where they enter this environment?

unset $(env | cut -d= -f1)


I suppose we should maybe *always* do that (first) since the user might be bringing in their own extra stuff when they source `env.sh`.

Reply:
good point, fixed in a2b9e06